### ISSUE TITLE
Value with querying multiple columns

### ISSUE THREAD

**Issue by sstok** [rollerworks/search-doctrine-dbal#9]
Say you have a value money "EUR 50.00", because of this format you can't use comparisons or ranges (_unless you are using using PostgreSQL with custom operators :metal:_).

You can choose to simply ignore the currency part and cast the remaining part to a numeric-type.
But if you _want_ to include the currency part you need to split the value to performs a comparison like `(SUBSTRING(price FROM 1 FOR 3) = 'EUR' AND CAST(SUBSTRING(price FROM 4) AS NUMERIC(2, 0)) > CAST('50.00' AS NUMERIC(2, 0)))` but this is currently not possible.

So we need a way to allow complete SQL fragment injection when using conversions. Any idea's?

**Comment by sstok** [comment]
Good news I have an idea.

`funcName($value, $... SubConditionCollector $condition)`. The `SubConditionCollector` (better name needed) allows to add new conditions (of any type) which will be evaluated by the query-generator. Basically its a subgroup-condition at query level, *but without input transforming and optimizing*.

The subcondition replaces the field (column) and value, what is set by the subcondition is transformed to a query (*the same way the original condition is transformed.*)

For the example above you something like: 

```
funcName($column, $value, ...., $condition) {
   $field = new ....($column);
   $field->addSingleValue($value) // method-names may differ

   $condition->...($field);
}
```

This implementation allows for more complex queries and hides the Query details with a user-friendly API.

**Comment by sstok** [comment]
I thought of multiple ways, and the best idea that came to my mind is this: a value-processor which generates a sub-condition (similar to a SearchCondition but less detailed)`[1]` which in turn translates to a sub-query.

The sub-query is applied secondary to the current condition. So if you have a range, the range value is still generated as SQL and the sub-query is applied within the sub-query-group as the range. For the example above you get:

```sql
(
    CAST(SUBSTRING(price FROM 4) AS NUMERIC(2, 0)) > CAST('50.00' AS NUMERIC(2, 0))
AND
    CAST(SUBSTRING(price FROM 4) AS NUMERIC(2, 0)) < CAST('80.00' AS NUMERIC(2, 0))
AND
    (SUBSTRING(price FROM 1 FOR 3) = 'EUR')
)
```

*In practice this will work the as if the currency had its own search-field and was applied within an AND-group.*

The sub-condition is explicitly not a SearchCondition, but does follow the same logic and structure.
Which means you can also add subgroups within the condition you are already in.
An important detail is that there is shared configuration between conversions, like we have with the FieldSet and WhereBuilder columns configuration.

Using conversions is still possible, but the sub-condition fields (columns) may have there own conversions. *Note that each field must have the conversions set explicitly and no global configuration is used.*

In other words the value-processor works as a subsystem for creating conditions which are applied directly within the current condition value.

`[1]`: Its also to use the Doctrine DBAL QueryBuilder, but this will be a problem for the Doctrine ORM processor which uses a mid-way system for executing SQL conversions. So unless we require the implementor to register any custom DQL that are used within the sub-condition this is not possible.
Personally I don't have a problem with using the QueryBuilder and registering custom DQL functions (whenever needed), as that is really big edge-case. And we can always add an alternative inteface which doesn't use the QueryBuilder (allowing for the best of both worlds).

**Comment by sstok** [comment]
Hmm, as the ORM and DBAL QeuryBuilder both have there own class its not possible to share there usage. So using a Condition class it our only option, unless we need to use a really really ugly hack with `(CASE WHEN (SUB-QUERY) THEN 1 ELSE 0 END) = 1` which is not good for performance and is just ugly :neutral_face:

**Comment by sstok** [comment]
Use QueryField for passing information to a field. Extract the QueryGenerator to split handling of value-types and query-structure.

**Comment by sstok** [comment]
Unless someone can help with this I'm postponing this until after 2.0 as I can't remember exactly what I planned 😰💀

**Comment by sstok** [comment]
I finally came-up with a suitable solution! It will be more work for the developer, but that's the price of flexibility 🤷‍♂️

Introduce a new `QueryStructureConversion` interface, which cannot be used in combination with other conversion types. When a Field has this conversion type configured both the field and value (holder) will be passed to the conversion. The Conversion MUST produce the full SQL for this field value:

```php
    public function getSql(string $column, /* ValueHolder|mixed */ $value, array $options, ConversionHints $hints): string
    {
        if ($value instanceof Range) {
            $currency = (string) $value->getLower()->value->getCurrency();
            
            return <<<SQL
(
    SUBSTRING(price FROM 1 FOR 3) = {$currency}
AND
    CAST(SUBSTRING(price FROM 4) AS NUMERIC(2, 0)) BETWEEN CAST({$hints->connection->quote($value->getLower()->value->getAmount())} AS NUMERIC(2, 0)) AND CAST({$hints->connection->quote($value->getUpper()->value->getAmount())} AS NUMERIC(2, 0))
)
SQL;
        }

        // ...
    }
```

Of course this will require more work for platform specific SQL but again that's the price of flexibility. The QueryPlatform will be made available using the `ConversionHints`. And maybe I can make this logic configure for easy re-usability.

**Comment by sstok** [comment]
Moved to https://github.com/rollerworks/search/issues/208
